{"cells":[{"cell_type":"markdown","metadata":{},"source":[" # Allen Brain Atlas Developing Human Brain"]},{"cell_type":"markdown","metadata":{},"source":[" Author: Moshe Silverstein <br/>\n"," Date: 3-18 <br/>\n"," Data Source: http://www.brainspan.org/static/download.html\n","\n"," Reviewer: Charles Dai <br>\n"," Updated: 6-20"]},{"cell_type":"code","execution_count":1,"metadata":{},"outputs":[],"source":["# appyter init\n","from appyter import magic\n","magic.init(lambda _=globals: _())"]},{"cell_type":"code","execution_count":2,"metadata":{},"outputs":[],"source":["import sys\n","import os\n","\n","import numpy as np\n","import pandas as pd\n","import matplotlib.pyplot as plt\n","%matplotlib inline\n","\n","import harmonizome-helper.utility_functions as uf\n","import harmonizome-helper.lookup as lookup"]},{"cell_type":"code","execution_count":3,"metadata":{},"outputs":[],"source":["# from clustergrammer_widget import *\n","# net = Network(clustergrammer_widget)"]},{"cell_type":"code","execution_count":4,"metadata":{},"outputs":[],"source":["%load_ext autoreload\n","%autoreload 2"]},{"cell_type":"markdown","metadata":{},"source":[" ### Python Version"]},{"cell_type":"code","execution_count":5,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"'3.8.0 (default, Oct 28 2019, 16:14:01) \\n[GCC 8.3.0]'"},"metadata":{},"execution_count":5}],"source":["sys.version"]},{"cell_type":"markdown","metadata":{},"source":[" # Initialization"]},{"cell_type":"markdown","metadata":{},"source":["### Set Data Grouping Options"]},{"cell_type":"code","execution_count":25,"metadata":{},"outputs":[],"source":["%%appyter code_eval\n","{% set data_type = ChoiceField(\n","    name='data_type',\n","    label='Data Type',\n","    choices={\n","        'RNA-Seq': 'RS', \n","        'Microarray': 'MA',\n","    },\n","    default='RNA-Seq',\n","    section='data'\n",") %}\n","\n","{% set data_grouping = ChoiceField(\n","    name='data_grouping',\n","    label='Data Grouped by',\n","    choices={\n","        'Age': \"'age'\", \n","        'Sample': \"'sample'\", \n","        'Tissue': \"'tissue'\"\n","    },\n","    default='Age',\n","    section='data'\n",") %}"]},{"cell_type":"code","execution_count":7,"metadata":{},"outputs":[{"output_type":"display_data","data":{"text/plain":"<IPython.core.display.Markdown object>","text/markdown":"```python\noptions = {\n    'age': {\n        'output_name': 'aba_dhb_RS_age',\n        'path': 'ABA-DHB-RS-A',\n        'grouping': lambda mat: mat['age']\n    },\n    'sample': {\n        'output_name': 'aba_dhb_RS_samp',\n        'path': 'ABA-DHB-RS-S',\n        'grouping':\n            lambda mat: mat.apply(\n                lambda row: '_'.join([\n                    row['structure_name'],\n                    str(row['age']),\n                    row['gender'],\n                    str(row['donor_id'])\n                ]), axis=1)\n    },\n    'tissue': {\n        'output_name': 'aba_dhb_RS_tissue',\n        'path': 'ABA-DHB-RS-T',\n        'grouping': lambda mat: mat['tissue']\n    }\n}\n```"},"metadata":{}}],"source":["%%appyter code_exec\n","\n","options = {\n","    'age': {\n","        'output_name': 'aba_dhb_{{data_type}}_age',\n","        'path': 'ABA-DHB-{{data_type}}-A',\n","        'grouping': lambda mat: mat['age']\n","    },\n","    'sample': {\n","        'output_name': 'aba_dhb_{{data_type}}_samp',\n","        'path': 'ABA-DHB-{{data_type}}-S',\n","        'grouping': \n","            lambda mat: mat.apply(\n","                lambda row: '_'.join([\n","                    row['structure_name'], \n","                    str(row['age']), \n","                    row['gender'], \n","                    str(row['donor_id'])\n","                ]), axis=1)\n","    },\n","    'tissue': {\n","        'output_name': 'aba_dhb_{{data_type}}_tissue',\n","        'path': 'ABA-DHB-{{data_type}}-T',\n","        'grouping': lambda mat: mat['tissue']\n","\n","    }\n","}"]},{"cell_type":"markdown","metadata":{},"source":[" ### Load Mapping Dictionaries"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["symbol_lookup, geneid_lookup = lookup.get_lookups()"]},{"cell_type":"markdown","metadata":{},"source":[" ### Output Path"]},{"cell_type":"code","execution_count":36,"metadata":{},"outputs":[{"output_type":"display_data","data":{"text/plain":"<IPython.core.display.Markdown object>","text/markdown":"```python\noutput_name = options['age']['output_name']\npath = options['age']['path']\nif not os.path.exists(path):\n    os.makedirs(path)\n```"},"metadata":{}}],"source":["%%appyter code_exec\n","\n","output_name = options[{{data_grouping}}]['output_name']\n","\n","path = options[{{data_grouping}}]['path']\n","if not os.path.exists(path):\n","    os.makedirs(path)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%appyter hide_code\n","{% do SectionField(\n","    name='data',\n","    title='Load Data',\n","    subtitle='Upload Files from Developing Human Brain Data Sets',\n",") %}"]},{"cell_type":"markdown","metadata":{},"source":[" # Load Data"]},{"cell_type":"code","execution_count":24,"metadata":{},"outputs":[{"output_type":"display_data","data":{"text/plain":"<IPython.core.display.Markdown object>","text/markdown":"```python\n\nmatrix = pd.read_csv('Input/ABA-DHB-MA/gene_array_matrix_csv/expression_matrix.csv', header=None, index_col=0)\n```"},"metadata":{}}],"source":["%%appyter code_exec\n","\n","matrix = pd.read_csv({{FileField(\n","    constraint='.*\\.csv$',\n","    name='expression_matrix', \n","    label='Expression Matrix', \n","    default='Input/ABA-DHB-MA/gene_array_matrix_csv/expression_matrix.csv',\n","    section='data')\n","}}, header=None, index_col=0)"]},{"cell_type":"code","execution_count":38,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"4431  6.62531  6.96386  6.74376  6.67833   \n5  4.57404  5.22911  4.66054  5.03038  5.02629  4.77163  4.89984  5.13641   \n\n       9        10   ...      483      484      485      486      487  \\\n0                    ...                                                \n1  9.16732  9.12500  ...  9.07099  8.77870  8.85986  8.34032  9.35445   \n2  8.54096  8.34537  ...  8.34760  8.45158  8.50765  8.19684  8.14229   \n3  9.55960  9.36183  ...  8.43173  8.54623  9.02475  8.21754  8.91127   \n4  7.06941  5.86169  ...  6.82257  7.54365  7.09902  6.71188  7.16314   \n5  5.27419  4.73643  ...  4.53204  4.72392  4.04599  5.12808  4.59776   \n\n       488      489      490      491      492  \n0                                               \n1  8.70370  9.64959  8.52157  9.22952  9.64878  \n2  8.54619  8.15847  8.77172  8.64942  8.18061  \n3  8.77168  9.11248  8.40019  9.05908  8.82711  \n4  7.18352  7.36694  6.70071  7.54916  6.78094  \n5  4.47878  4.37817  4.59131  4.96854  4.35579  \n\n[5 rows x 492 columns]","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>1</th>\n      <th>2</th>\n      <th>3</th>\n      <th>4</th>\n      <th>5</th>\n      <th>6</th>\n      <th>7</th>\n      <th>8</th>\n      <th>9</th>\n      <th>10</th>\n      <th>...</th>\n      <th>483</th>\n      <th>484</th>\n      <th>485</th>\n      <th>486</th>\n      <th>487</th>\n      <th>488</th>\n      <th>489</th>\n      <th>490</th>\n      <th>491</th>\n      <th>492</th>\n    </tr>\n    <tr>\n      <th>0</th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>1</th>\n      <td>9.16596</td>\n      <td>9.35303</td>\n      <td>9.12842</td>\n      <td>9.28837</td>\n      <td>9.17170</td>\n      <td>9.52198</td>\n      <td>9.26660</td>\n      <td>9.31059</td>\n      <td>9.16732</td>\n      <td>9.12500</td>\n      <td>...</td>\n      <td>9.07099</td>\n      <td>8.77870</td>\n      <td>8.85986</td>\n      <td>8.34032</td>\n      <td>9.35445</td>\n      <td>8.70370</td>\n      <td>9.64959</td>\n      <td>8.52157</td>\n      <td>9.22952</td>\n      <td>9.64878</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>8.40401</td>\n      <td>8.43084</td>\n      <td>8.87691</td>\n      <td>8.81414</td>\n      <td>8.83160</td>\n      <td>8.56038</td>\n      <td>8.42721</td>\n      <td>8.85843</td>\n      <td>8.54096</td>\n      <td>8.34537</td>\n      <td>...</td>\n      <td>8.34760</td>\n      <td>8.45158</td>\n      <td>8.50765</td>\n      <td>8.19684</td>\n      <td>8.14229</td>\n      <td>8.54619</td>\n      <td>8.15847</td>\n      <td>8.77172</td>\n      <td>8.64942</td>\n      <td>8.18061</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>9.35942</td>\n      <td>9.58458</td>\n      <td>9.76375</td>\n      <td>9.53778</td>\n      <td>9.61363</td>\n      <td>9.37657</td>\n      <td>9.55275</td>\n      <td>9.83603</td>\n      <td>9.55960</td>\n      <td>9.36183</td>\n      <td>...</td>\n      <td>8.43173</td>\n      <td>8.54623</td>\n      <td>9.02475</td>\n      <td>8.21754</td>\n      <td>8.91127</td>\n      <td>8.77168</td>\n      <td>9.11248</td>\n      <td>8.40019</td>\n      <td>9.05908</td>\n      <td>8.82711</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>6.30357</td>\n      <td>6.73438</td>\n      <td>6.50582</td>\n      <td>7.24431</td>\n      <td>6.62531</td>\n      <td>6.96386</td>\n      <td>6.74376</td>\n      <td>6.67833</td>\n      <td>7.06941</td>\n      <td>5.86169</td>\n      <td>...</td>\n      <td>6.82257</td>\n      <td>7.54365</td>\n      <td>7.09902</td>\n      <td>6.71188</td>\n      <td>7.16314</td>\n      <td>7.18352</td>\n      <td>7.36694</td>\n      <td>6.70071</td>\n      <td>7.54916</td>\n      <td>6.78094</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>4.57404</td>\n      <td>5.22911</td>\n      <td>4.66054</td>\n      <td>5.03038</td>\n      <td>5.02629</td>\n      <td>4.77163</td>\n      <td>4.89984</td>\n      <td>5.13641</td>\n      <td>5.27419</td>\n      <td>4.73643</td>\n      <td>...</td>\n      <td>4.53204</td>\n      <td>4.72392</td>\n      <td>4.04599</td>\n      <td>5.12808</td>\n      <td>4.59776</td>\n      <td>4.47878</td>\n      <td>4.37817</td>\n      <td>4.59131</td>\n      <td>4.96854</td>\n      <td>4.35579</td>\n    </tr>\n  </tbody>\n</table>\n<p>5 rows × 492 columns</p>\n</div>"},"metadata":{},"execution_count":38}],"source":["matrix.head()"]},{"cell_type":"code","execution_count":39,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"(17604, 492)"},"metadata":{},"execution_count":39}],"source":["matrix.shape"]},{"cell_type":"markdown","metadata":{},"source":[" ## Load Sample Metadata"]},{"cell_type":"code","execution_count":40,"metadata":{},"outputs":[{"output_type":"display_data","data":{"text/plain":"<IPython.core.display.Markdown object>","text/markdown":"```python\n\nsample_meta = pd.read_csv('Input/ABA-DHB-MA-S/gene_array_matrix_csv/columns_metadata.csv', index_col=0)\n```"},"metadata":{}}],"source":["%%appyter code_exec\n","\n","sample_meta = pd.read_csv({{FileField(\n","    constraint='.*\\.csv$',\n","    name='columns_metadata', \n","    label='Sample Metadata', \n","    default='Input/ABA-DHB-{{data_type}}/gene_array_matrix_csv/columns_metadata.csv',\n","    section='data')\n","}}, index_col=0)"]},{"cell_type":"code","execution_count":41,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"            donor_id   donor_name    age gender  structure_id  \\\ncolumn_num                                                      \n1              13058  H376.IIA.51  8 pcw      M         10361   \n2              13058  H376.IIA.51  8 pcw      M         10552   \n3              13058  H376.IIA.51  8 pcw      M         10173   \n4              13058  H376.IIA.51  8 pcw      M         10391   \n5              13058  H376.IIA.51  8 pcw      M         10294   \n\n           structure_acronym                       structure_name  \ncolumn_num                                                         \n1                        AMY                   amygdaloid complex  \n2                        CGE           caudal ganglionic eminence  \n3                        DFC       dorsolateral prefrontal cortex  \n4                        DTH                      dorsal thalamus  \n5                        HIP  hippocampus (hippocampal formation)  ","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>donor_id</th>\n      <th>donor_name</th>\n      <th>age</th>\n      <th>gender</th>\n      <th>structure_id</th>\n      <th>structure_acronym</th>\n      <th>structure_name</th>\n    </tr>\n    <tr>\n      <th>column_num</th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>1</th>\n      <td>13058</td>\n      <td>H376.IIA.51</td>\n      <td>8 pcw</td>\n      <td>M</td>\n      <td>10361</td>\n      <td>AMY</td>\n      <td>amygdaloid complex</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>13058</td>\n      <td>H376.IIA.51</td>\n      <td>8 pcw</td>\n      <td>M</td>\n      <td>10552</td>\n      <td>CGE</td>\n      <td>caudal ganglionic eminence</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>13058</td>\n      <td>H376.IIA.51</td>\n      <td>8 pcw</td>\n      <td>M</td>\n      <td>10173</td>\n      <td>DFC</td>\n      <td>dorsolateral prefrontal cortex</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>13058</td>\n      <td>H376.IIA.51</td>\n      <td>8 pcw</td>\n      <td>M</td>\n      <td>10391</td>\n      <td>DTH</td>\n      <td>dorsal thalamus</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>13058</td>\n      <td>H376.IIA.51</td>\n      <td>8 pcw</td>\n      <td>M</td>\n      <td>10294</td>\n      <td>HIP</td>\n      <td>hippocampus (hippocampal formation)</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{},"execution_count":41}],"source":["sample_meta.head()"]},{"cell_type":"code","execution_count":42,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"(492, 7)"},"metadata":{},"execution_count":42}],"source":["sample_meta.shape"]},{"cell_type":"markdown","metadata":{},"source":[" ## Load Gene Metadata"]},{"cell_type":"code","execution_count":43,"metadata":{},"outputs":[{"output_type":"display_data","data":{"text/plain":"<IPython.core.display.Markdown object>","text/markdown":"```python\n\ngene_meta = pd.read_csv('Input/ABA-DHB-MA-S/gene_array_matrix_csv/rows_metadata.csv', index_col=0)\n```"},"metadata":{}}],"source":["%%appyter code_exec\n","\n","gene_meta = pd.read_csv({{FileField(\n","    constraint='.*\\.csv$',\n","    name='rows_metadata', \n","    label='Gene Metadata', \n","    default='Input/ABA-DHB-{{data_type}}/gene_array_matrix_csv/rows_metadata.csv',\n","    section='data')\n","}}, index_col=0)"]},{"cell_type":"code","execution_count":44,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"         gene_id  ensembl_gene_id gene_symbol  entrez_id\nrow_num                                                 \n1           9633  ENSG00000101337      TM9SF4       9777\n2          22692  ENSG00000114650        SCAP      22937\n3          22952  ENSG00000113194        FAF2      23197\n4          34497  ENSG00000150401     DCUN1D2      55208\n5          35224  ENSG00000204444        APOM      55937","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>gene_id</th>\n      <th>ensembl_gene_id</th>\n      <th>gene_symbol</th>\n      <th>entrez_id</th>\n    </tr>\n    <tr>\n      <th>row_num</th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>1</th>\n      <td>9633</td>\n      <td>ENSG00000101337</td>\n      <td>TM9SF4</td>\n      <td>9777</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>22692</td>\n      <td>ENSG00000114650</td>\n      <td>SCAP</td>\n      <td>22937</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>22952</td>\n      <td>ENSG00000113194</td>\n      <td>FAF2</td>\n      <td>23197</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>34497</td>\n      <td>ENSG00000150401</td>\n      <td>DCUN1D2</td>\n      <td>55208</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>35224</td>\n      <td>ENSG00000204444</td>\n      <td>APOM</td>\n      <td>55937</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{},"execution_count":44}],"source":["gene_meta.head()"]},{"cell_type":"code","execution_count":45,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"(17604, 4)"},"metadata":{},"execution_count":45}],"source":["gene_meta.shape"]},{"cell_type":"markdown","metadata":{},"source":[" # Pre-process Data"]},{"cell_type":"markdown","metadata":{},"source":[" ## Map Sample Metadata to Sample ID"]},{"cell_type":"code","execution_count":46,"metadata":{},"outputs":[{"output_type":"display_data","data":{"text/plain":"<IPython.core.display.Markdown object>","text/markdown":"```python\nmap_func = options['age']['grouping']\nmatrix.columns = map_func(sample_meta)\nmatrix.head()\n```"},"metadata":{}}],"source":["%%appyter code_exec\n","\n","map_func = options[{{data_grouping}}]['grouping']\n","matrix.columns = map_func(sample_meta)\n","matrix.head()"]},{"cell_type":"markdown","metadata":{},"source":[" ## Map Gene to Row"]},{"cell_type":"code","execution_count":47,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"582  7.24431  6.62531  6.96386  6.74376   \nAPOM     4.57404  5.22911  4.66054  5.03038  5.02629  4.77163  4.89984   \n\nage        8 pcw    8 pcw    8 pcw  ...   40 yrs   40 yrs   40 yrs   40 yrs  \\\nTM9SF4   9.31059  9.16732  9.12500  ...  9.07099  8.77870  8.85986  8.34032   \nSCAP     8.85843  8.54096  8.34537  ...  8.34760  8.45158  8.50765  8.19684   \nFAF2     9.83603  9.55960  9.36183  ...  8.43173  8.54623  9.02475  8.21754   \nDCUN1D2  6.67833  7.06941  5.86169  ...  6.82257  7.54365  7.09902  6.71188   \nAPOM     5.13641  5.27419  4.73643  ...  4.53204  4.72392  4.04599  5.12808   \n\nage       40 yrs   40 yrs   40 yrs   40 yrs   40 yrs   40 yrs  \nTM9SF4   9.35445  8.70370  9.64959  8.52157  9.22952  9.64878  \nSCAP     8.14229  8.54619  8.15847  8.77172  8.64942  8.18061  \nFAF2     8.91127  8.77168  9.11248  8.40019  9.05908  8.82711  \nDCUN1D2  7.16314  7.18352  7.36694  6.70071  7.54916  6.78094  \nAPOM     4.59776  4.47878  4.37817  4.59131  4.96854  4.35579  \n\n[5 rows x 492 columns]","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th>age</th>\n      <th>8 pcw</th>\n      <th>8 pcw</th>\n      <th>8 pcw</th>\n      <th>8 pcw</th>\n      <th>8 pcw</th>\n      <th>8 pcw</th>\n      <th>8 pcw</th>\n      <th>8 pcw</th>\n      <th>8 pcw</th>\n      <th>8 pcw</th>\n      <th>...</th>\n      <th>40 yrs</th>\n      <th>40 yrs</th>\n      <th>40 yrs</th>\n      <th>40 yrs</th>\n      <th>40 yrs</th>\n      <th>40 yrs</th>\n      <th>40 yrs</th>\n      <th>40 yrs</th>\n      <th>40 yrs</th>\n      <th>40 yrs</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>TM9SF4</th>\n      <td>9.16596</td>\n      <td>9.35303</td>\n      <td>9.12842</td>\n      <td>9.28837</td>\n      <td>9.17170</td>\n      <td>9.52198</td>\n      <td>9.26660</td>\n      <td>9.31059</td>\n      <td>9.16732</td>\n      <td>9.12500</td>\n      <td>...</td>\n      <td>9.07099</td>\n      <td>8.77870</td>\n      <td>8.85986</td>\n      <td>8.34032</td>\n      <td>9.35445</td>\n      <td>8.70370</td>\n      <td>9.64959</td>\n      <td>8.52157</td>\n      <td>9.22952</td>\n      <td>9.64878</td>\n    </tr>\n    <tr>\n      <th>SCAP</th>\n      <td>8.40401</td>\n      <td>8.43084</td>\n      <td>8.87691</td>\n      <td>8.81414</td>\n      <td>8.83160</td>\n      <td>8.56038</td>\n      <td>8.42721</td>\n      <td>8.85843</td>\n      <td>8.54096</td>\n      <td>8.34537</td>\n      <td>...</td>\n      <td>8.34760</td>\n      <td>8.45158</td>\n      <td>8.50765</td>\n      <td>8.19684</td>\n      <td>8.14229</td>\n      <td>8.54619</td>\n      <td>8.15847</td>\n      <td>8.77172</td>\n      <td>8.64942</td>\n      <td>8.18061</td>\n    </tr>\n    <tr>\n      <th>FAF2</th>\n      <td>9.35942</td>\n      <td>9.58458</td>\n      <td>9.76375</td>\n      <td>9.53778</td>\n      <td>9.61363</td>\n      <td>9.37657</td>\n      <td>9.55275</td>\n      <td>9.83603</td>\n      <td>9.55960</td>\n      <td>9.36183</td>\n      <td>...</td>\n      <td>8.43173</td>\n      <td>8.54623</td>\n      <td>9.02475</td>\n      <td>8.21754</td>\n      <td>8.91127</td>\n      <td>8.77168</td>\n      <td>9.11248</td>\n      <td>8.40019</td>\n      <td>9.05908</td>\n      <td>8.82711</td>\n    </tr>\n    <tr>\n      <th>DCUN1D2</th>\n      <td>6.30357</td>\n      <td>6.73438</td>\n      <td>6.50582</td>\n      <td>7.24431</td>\n      <td>6.62531</td>\n      <td>6.96386</td>\n      <td>6.74376</td>\n      <td>6.67833</td>\n      <td>7.06941</td>\n      <td>5.86169</td>\n      <td>...</td>\n      <td>6.82257</td>\n      <td>7.54365</td>\n      <td>7.09902</td>\n      <td>6.71188</td>\n      <td>7.16314</td>\n      <td>7.18352</td>\n      <td>7.36694</td>\n      <td>6.70071</td>\n      <td>7.54916</td>\n      <td>6.78094</td>\n    </tr>\n    <tr>\n      <th>APOM</th>\n      <td>4.57404</td>\n      <td>5.22911</td>\n      <td>4.66054</td>\n      <td>5.03038</td>\n      <td>5.02629</td>\n      <td>4.77163</td>\n      <td>4.89984</td>\n      <td>5.13641</td>\n      <td>5.27419</td>\n      <td>4.73643</td>\n      <td>...</td>\n      <td>4.53204</td>\n      <td>4.72392</td>\n      <td>4.04599</td>\n      <td>5.12808</td>\n      <td>4.59776</td>\n      <td>4.47878</td>\n      <td>4.37817</td>\n      <td>4.59131</td>\n      <td>4.96854</td>\n      <td>4.35579</td>\n    </tr>\n  </tbody>\n</table>\n<p>5 rows × 492 columns</p>\n</div>"},"metadata":{},"execution_count":47}],"source":["matrix.index = gene_meta['gene_symbol']\n","matrix.index.name = None\n","matrix.head()"]},{"cell_type":"code","execution_count":48,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"(17604, 492)"},"metadata":{},"execution_count":48}],"source":["matrix.shape"]},{"cell_type":"markdown","metadata":{},"source":[" ## Save Unfiltered Matrix to file"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(matrix, path, output_name + '_matrix_unfiltered',\n","            compression='gzip', dtype=np.float32)"]},{"cell_type":"markdown","metadata":{},"source":[" # Filter Data"]},{"cell_type":"markdown","metadata":{},"source":[" ## Remove Data that is More Than 95% Missing and Impute Missing Data"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix = uf.removeAndImpute(matrix)\n","matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix.shape"]},{"cell_type":"markdown","metadata":{},"source":[" ## Map Gene Symbols to Up-to-date Approved Gene Symbols"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix = uf.mapgenesymbols(matrix, symbol_lookup)\n","matrix.shape"]},{"cell_type":"markdown","metadata":{},"source":[" ## Merge Duplicate Genes By Rows and Duplicate Columns"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix = uf.merge(matrix, 'row', 'mean')\n","matrix = uf.merge(matrix, 'column', 'mean')\n","matrix.shape"]},{"cell_type":"markdown","metadata":{},"source":[" ## Log2 Transform"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix = uf.log2(matrix)\n","matrix.head()"]},{"cell_type":"markdown","metadata":{},"source":[" ## Normalize Matrix (Quantile Normalize the Matrix by Column)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix = uf.quantileNormalize(matrix)\n","matrix.head()"]},{"cell_type":"markdown","metadata":{},"source":[" ## Normalize Matrix (Z-Score the Rows)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix = uf.zscore(matrix, 'row')\n","matrix.head()"]},{"cell_type":"markdown","metadata":{},"source":[" ## Histogram of First Sample"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix.iloc[:, 0].hist(bins=100)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Histogram of First Gene"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix.iloc[0, :].hist(bins=100)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Save Filtered Matrix"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(matrix, path, output_name + '_matrix_filtered', \n","            ext='tsv', compression='gzip')"]},{"cell_type":"markdown","metadata":{},"source":[" # Analyze Data"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Gene List"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["gene_list = uf.createGeneList(matrix, geneid_lookup)\n","gene_list.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["gene_list.shape"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(gene_list, path, output_name + '_gene_list',\n","            ext='tsv', compression='gzip', index=False)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Attribute List"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["attribute_list = uf.createAttributeList(matrix)\n","attribute_list.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["attribute_list.shape"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(attribute_list, path, output_name + '_attribute_list',\n","            ext='tsv', compression='gzip')"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create matrix of Standardized values (values between -1, and 1)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["standard_matrix = uf.createStandardizedMatrix(matrix)\n","standard_matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(standard_matrix, path, output_name + '_standard_matrix',\n","            ext='tsv', compression='gzip')"]},{"cell_type":"markdown","metadata":{},"source":[" ## Plot of A Single Celltype, Normalized Value vs. Standardized Value"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["plt.plot(matrix[matrix.columns[0]],\n","         standard_matrix[standard_matrix.columns[0]], 'bo')\n","plt.xlabel('Normalized Values')\n","plt.ylabel('Standardized Values')\n","plt.title(standard_matrix.columns[0])\n","plt.grid(True)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Ternary Matrix"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["ternary_matrix = uf.createTernaryMatrix(standard_matrix)\n","ternary_matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(ternary_matrix, path, output_name + '_ternary_matrix',\n","            ext='tsv', compression='gzip')"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Gene and Attribute Set Libraries"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.createUpGeneSetLib(ternary_matrix, path, output_name + '_gene_up_set')"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.createDownGeneSetLib(ternary_matrix, path, output_name + '_gene_down_set')"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.createUpAttributeSetLib(ternary_matrix, path, \n","                           output_name + '_attribute_up_set')"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.createDownAttributeSetLib(ternary_matrix, path, \n","                             output_name + '_attribute_down_set')"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Attribute Similarity Matrix"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["attribute_similarity_matrix = uf.createSimilarityMatrix(matrix.T, 'cosine')\n","attribute_similarity_matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(attribute_similarity_matrix, path,\n","            output_name + '_attribute_similarity_matrix', \n","            compression='npz', symmetric=True, dtype=np.float32)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# net.load_df(attribute_similarity_matrix.iloc[:,:].copy())\n","# net.filter_N_top('row', rank_type='sum', N_top=300)\n","# net.cluster()\n","# net.widget()"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Gene Similarity Matrix"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["gene_similarity_matrix = uf.createSimilarityMatrix(matrix, 'cosine')\n","gene_similarity_matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(gene_similarity_matrix, path, \n","            output_name + '_gene_similarity_matrix',\n","            compression='npz', symmetric=True, dtype=np.float32)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Gene-Attribute Edge List"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.createGeneAttributeEdgeList(standard_matrix, attribute_list, gene_list, \n","                               path, output_name + '_gene_attribute_edge_list')"]},{"cell_type":"markdown","metadata":{},"source":[" # Create Downloadable Save File"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.createArchive(path)"]},{"cell_type":"markdown","metadata":{},"source":[" ### Link to download output files: [click here](./output_archive.zip)"]}],"metadata":{"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.0-final"},"orig_nbformat":2,"kernelspec":{"name":"python38064bitharmonizomevenve1fd62c035f347dbaffce382c9034785","display_name":"Python 3.8.0 64-bit ('harmonizome': venv)"}},"nbformat":4,"nbformat_minor":2}