{"cells":[{"cell_type":"markdown","metadata":{},"source":[" # GTEx Tissue"]},{"cell_type":"markdown","metadata":{},"source":[" Author: Moshe Silverstein <br/>\n"," Date: 12-2017 <br/>\n"," Data Source: https://www.gtexportal.org/home/datasets\n","\n"," Reviewer: Charles Dai <br>\n"," Updated: 6-20"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# appyter init\n","from appyter import magic\n","magic.init(lambda _=globals: _())"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["import sys\n","import os\n","import gzip\n","\n","import numpy as np\n","import pandas as pd\n","import matplotlib.pyplot as plt\n","%matplotlib inline\n","\n","import utility_functions as uf\n","import lookup"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# from clustergrammer_widget import *\n","# net = Network(clustergrammer_widget)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%load_ext autoreload\n","%autoreload 2"]},{"cell_type":"markdown","metadata":{},"source":[" ### Python Version"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["sys.version"]},{"cell_type":"markdown","metadata":{},"source":[" # Initialization"]},{"cell_type":"markdown","metadata":{},"source":[" ### Load Mapping Dictionaries"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["symbol_lookup, geneid_lookup = lookup.get_lookups()"]},{"cell_type":"markdown","metadata":{},"source":[" ### Output Path"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["output_name = 'gtex'\n","\n","path = 'Output/GTEx'\n","if not os.path.exists(path):\n","    os.makedirs(path)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%appyter hide_code\n","{% do SectionField(\n","    name='data',\n","    title='Load Data',\n","    subtitle='Upload Files from the GTEx Portal',\n",") %}"]},{"cell_type":"markdown","metadata":{},"source":[" # Load Data"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%appyter code_exec\n","\n","matrix_zip = {{FileField(\n","    constraint='.*\\.gz$',\n","    name='expression_matrix', \n","    label='RNA-Seq Gene TPMs (gz)', \n","    default='Input/GTEx/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct.gz',\n","    section='data')\n","}}\n","with gzip.open(matrix_zip, 'r') as matrix_file:\n","    matrix = pd.read_csv(matrix_file, sep='\\t', skiprows=2)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix.shape"]},{"cell_type":"markdown","metadata":{},"source":[" ## Load Sample Metadata"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%appyter code_exec\n","\n","sample_meta = pd.read_csv({{FileField(\n","    constraint='.*\\.txt$',\n","    name='sample_metadata', \n","    label='Sample Metadata', \n","    default='Input/GTEx/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt',\n","    section='data')\n","}}, sep='\\t', index_col=0, usecols='SMTSD')"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["sample_meta.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["sample_meta.shape\n","'''"]},{"cell_type":"markdown","metadata":{},"source":[" ## Load Subject Metadata"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%appyter code_exec\n","\n","subject_meta = pd.read_csv({{FileField(\n","    constraint='.*\\.txt$',\n","    name='subject_metadata', \n","    label='Subject Metadata', \n","    default='Input/GTEx/GTEx_Analysis_v8_Annotations_SubjectPhenotypesDS.txt',\n","    section='data')\n","}})"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["subject_meta.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["subject_meta.shape\n","'''"]},{"cell_type":"markdown","metadata":{},"source":[" # Pre-process Data"]},{"cell_type":"markdown","metadata":{},"source":[" ## Map Sample ID to Attribute"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix.columns = sample_meta.reindex(matrix.columns).reset_index(drop=True)\n","matrix.head()"]},{"cell_type":"markdown","metadata":{},"source":[" ## Save Unfiltered Matrix to file"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(matrix, path, output_name + '_matrix_unfiltered',\n","            compression='gzip', dtype=np.float32)"]},{"cell_type":"markdown","metadata":{},"source":[" # Filter Data"]},{"cell_type":"markdown","metadata":{},"source":[" ## Remove Data that is More Than 95% Missing and Impute Missing Data"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix = uf.removeAndImpute(matrix)\n","matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix.shape"]},{"cell_type":"markdown","metadata":{},"source":[" ## Map Gene Symbols to Up-to-date Approved Gene Symbols"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix = uf.mapgenesymbols(matrix, symbol_lookup)\n","matrix.shape"]},{"cell_type":"markdown","metadata":{},"source":[" ## Merge Duplicate Genes By Rows and Duplicate Columns"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix = uf.merge(matrix, 'row', 'mean')\n","matrix = uf.merge(matrix, 'column', 'mean')\n","matrix.shape"]},{"cell_type":"markdown","metadata":{},"source":[" ## Log2 Transform"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix = uf.log2(matrix)\n","matrix.head()"]},{"cell_type":"markdown","metadata":{},"source":[" ## Normalize Matrix (Quantile Normalize the Matrix by Column)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix = uf.quantileNormalize(matrix)\n","matrix.head()"]},{"cell_type":"markdown","metadata":{},"source":[" ## Normalize Matrix (Z-Score the Rows)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix = uf.zscore(matrix, 'row')\n","matrix.head()"]},{"cell_type":"markdown","metadata":{},"source":[" ## Histogram of First Sample"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix.iloc[:, 0].hist(bins=100)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Histogram of First Gene"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["matrix.iloc[0, :].hist(bins=100)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Save Filtered Matrix"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(matrix, path, output_name + '_matrix_filtered', \n","            ext='tsv', compression='gzip')"]},{"cell_type":"markdown","metadata":{},"source":[" # Analyze Data"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Gene List"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["gene_list = uf.createGeneList(matrix, geneid_lookup)\n","gene_list.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["gene_list.shape"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(gene_list, path, output_name + '_gene_list',\n","            ext='tsv', compression='gzip', index=False)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Attribute List"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["attribute_list = uf.createAttributeList(matrix)\n","attribute_list.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["attribute_list.shape"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(attribute_list, path, output_name + '_attribute_list',\n","            ext='tsv', compression='gzip')"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create matrix of Standardized values (values between -1, and 1)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["standard_matrix = uf.createStandardizedMatrix(matrix)\n","standard_matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(standard_matrix, path, output_name + '_standard_matrix',\n","            ext='tsv', compression='gzip')"]},{"cell_type":"markdown","metadata":{},"source":[" ## Plot of A Single Celltype, Normalized Value vs. Standardized Value"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["plt.plot(matrix[matrix.columns[0]],\n","         standard_matrix[standard_matrix.columns[0]], 'bo')\n","plt.xlabel('Normalized Values')\n","plt.ylabel('Standardized Values')\n","plt.title(standard_matrix.columns[0])\n","plt.grid(True)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Ternary Matrix"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["ternary_matrix = uf.createTernaryMatrix(standard_matrix)\n","ternary_matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(ternary_matrix, path, output_name + '_ternary_matrix',\n","            ext='tsv', compression='gzip')"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Gene and Attribute Set Libraries"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.createUpGeneSetLib(ternary_matrix, path, output_name + '_gene_up_set')"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.createDownGeneSetLib(ternary_matrix, path, output_name + '_gene_down_set')"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.createUpAttributeSetLib(ternary_matrix, path, \n","                           output_name + '_attribute_up_set')"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.createDownAttributeSetLib(ternary_matrix, path, \n","                             output_name + '_attribute_down_set')"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Attribute Similarity Matrix"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["attribute_similarity_matrix = uf.createSimilarityMatrix(matrix.T, 'cosine')\n","attribute_similarity_matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(attribute_similarity_matrix, path,\n","            output_name + '_attribute_similarity_matrix', \n","            compression='npz', symmetric=True, dtype=np.float32)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# net.load_df(attribute_similarity_matrix.iloc[:,:].copy())\n","# net.filter_N_top('row', rank_type='sum', N_top=300)\n","# net.cluster()\n","# net.widget()"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Gene Similarity Matrix"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["gene_similarity_matrix = uf.createSimilarityMatrix(matrix, 'cosine')\n","gene_similarity_matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(gene_similarity_matrix, path, \n","            output_name + '_gene_similarity_matrix',\n","            compression='npz', symmetric=True, dtype=np.float32)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Gene-Attribute Edge List"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.createGeneAttributeEdgeList(standard_matrix, attribute_list, gene_list, \n","                               path, output_name + '_gene_attribute_edge_list')"]},{"cell_type":"markdown","metadata":{},"source":[" # Create Downloadable Save File"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.createArchive(path)"]},{"cell_type":"markdown","metadata":{},"source":[" ### Link to download output files: [click here](./output_archive.zip)"]}],"nbformat":4,"nbformat_minor":2,"metadata":{"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":3},"orig_nbformat":2}}