{"cells":[{"cell_type":"markdown","metadata":{},"source":[" # GWAS Catalog (SNP-Phenotype Associations)"]},{"cell_type":"markdown","metadata":{},"source":[" Author: Moshe Silverstein <br/>\n"," Date: 9-17 <br/>\n"," Data Source: http://www.ebi.ac.uk/gwas/docs/file-downloads\n","\n"," Reviewer: Charles Dai <br>\n"," Updated: 6-20"]},{"cell_type":"code","execution_count":1,"metadata":{},"outputs":[],"source":["# appyter init\n","from appyter import magic\n","magic.init(lambda _=globals: _())"]},{"cell_type":"code","execution_count":2,"metadata":{},"outputs":[],"source":["import sys\n","import os\n","import re\n","\n","import numpy as np\n","import pandas as pd\n","import matplotlib.pyplot as plt\n","%matplotlib inline\n","\n","import utility_functions as uf\n","import lookup"]},{"cell_type":"code","execution_count":3,"metadata":{},"outputs":[],"source":["# from clustergrammer_widget import *\n","# net = Network(clustergrammer_widget)"]},{"cell_type":"code","execution_count":4,"metadata":{},"outputs":[],"source":["%load_ext autoreload\n","%autoreload 2"]},{"cell_type":"markdown","metadata":{},"source":[" ### Python Version"]},{"cell_type":"code","execution_count":5,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"'3.8.0 (default, Oct 28 2019, 16:14:01) \\n[GCC 8.3.0]'"},"metadata":{},"execution_count":5}],"source":["sys.version"]},{"cell_type":"markdown","metadata":{},"source":[" # Initialization"]},{"cell_type":"markdown","metadata":{},"source":[" ### Load Mapping Dictionaries"]},{"cell_type":"code","execution_count":6,"metadata":{},"outputs":[{"output_type":"stream","name":"stderr","text":"Gathering sources: 100%|██████████| 3/3 [00:08<00:00,  2.92s/it]\n"}],"source":["symbol_lookup, geneid_lookup = lookup.get_lookups()"]},{"cell_type":"markdown","metadata":{},"source":[" ### Output Path"]},{"cell_type":"code","execution_count":7,"metadata":{},"outputs":[],"source":["output_name = 'gwas_catalog'\n","\n","path = 'Output/GWAS-CATALOG'\n","if not os.path.exists(path):\n","    os.makedirs(path)"]},{"cell_type":"code","execution_count":8,"metadata":{},"outputs":[],"source":["%%appyter hide_code\n","{% do SectionField(\n","    name='data',\n","    title='Load Data',\n","    subtitle='Upload files from the GWAS CAtalog Associations Data Set',\n",") %}"]},{"cell_type":"markdown","metadata":{},"source":[" # Load Data"]},{"cell_type":"code","execution_count":9,"metadata":{},"outputs":[{"output_type":"display_data","data":{"text/plain":"<IPython.core.display.Markdown object>","text/markdown":"```python\n\ndf = pd.read_csv('Input/GWAS-Catalog/gwas_catalog_v1.0.2-associations_e100_r2020-06-04.tsv', sep='\\t', usecols=['DISEASE/TRAIT', 'MAPPED_GENE'])\n```"},"metadata":{}}],"source":["%%appyter code_exec\n","\n","df = pd.read_csv({{FileField(\n","    constraint='.*\\.tsv$',\n","    name='associations', \n","    label='All Associations', \n","    default='Input/GWAS-Catalog/gwas_catalog_v1.0.2-associations_e100_r2020-06-04.tsv',\n","    section='data')\n","}}, sep='\\t', usecols=['DISEASE/TRAIT', 'MAPPED_GENE'])"]},{"cell_type":"code","execution_count":10,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"           DISEASE/TRAIT             MAPPED_GENE\n0  Sudden cardiac arrest     FTH1P5 - AL158050.2\n1  Sudden cardiac arrest  LINC02376 - AC073913.1\n2  Sudden cardiac arrest                  CHRNB4\n3  Sudden cardiac arrest    MARCHF10, AC005821.1\n4  Sudden cardiac arrest                    CDH4","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>DISEASE/TRAIT</th>\n      <th>MAPPED_GENE</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>Sudden cardiac arrest</td>\n      <td>FTH1P5 - AL158050.2</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>Sudden cardiac arrest</td>\n      <td>LINC02376 - AC073913.1</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>Sudden cardiac arrest</td>\n      <td>CHRNB4</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>Sudden cardiac arrest</td>\n      <td>MARCHF10, AC005821.1</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>Sudden cardiac arrest</td>\n      <td>CDH4</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{},"execution_count":10}],"source":["df.head()"]},{"cell_type":"code","execution_count":11,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"(186829, 2)"},"metadata":{},"execution_count":11}],"source":["df.shape"]},{"cell_type":"markdown","metadata":{},"source":[" # Pre-process Data"]},{"cell_type":"markdown","metadata":{},"source":[" ## Split Gene Lists"]},{"cell_type":"code","execution_count":12,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"           DISEASE/TRAIT              MAPPED_GENE\n0  Sudden cardiac arrest     [FTH1P5, AL158050.2]\n1  Sudden cardiac arrest  [LINC02376, AC073913.1]\n2  Sudden cardiac arrest                 [CHRNB4]\n3  Sudden cardiac arrest   [MARCHF10, AC005821.1]\n4  Sudden cardiac arrest                   [CDH4]","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>DISEASE/TRAIT</th>\n      <th>MAPPED_GENE</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>Sudden cardiac arrest</td>\n      <td>[FTH1P5, AL158050.2]</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>Sudden cardiac arrest</td>\n      <td>[LINC02376, AC073913.1]</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>Sudden cardiac arrest</td>\n      <td>[CHRNB4]</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>Sudden cardiac arrest</td>\n      <td>[MARCHF10, AC005821.1]</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>Sudden cardiac arrest</td>\n      <td>[CDH4]</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{},"execution_count":12}],"source":["df = df.dropna()\n","df['MAPPED_GENE'] = df['MAPPED_GENE'].str.split(pat= '; | - ')\n","df.head()"]},{"cell_type":"code","execution_count":13,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"                     DISEASE/TRAIT\nMAPPED_GENE                       \nFTH1P5       Sudden cardiac arrest\nAL158050.2   Sudden cardiac arrest\nLINC02376    Sudden cardiac arrest\nAC073913.1   Sudden cardiac arrest\nCHRNB4       Sudden cardiac arrest","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>DISEASE/TRAIT</th>\n    </tr>\n    <tr>\n      <th>MAPPED_GENE</th>\n      <th></th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>FTH1P5</th>\n      <td>Sudden cardiac arrest</td>\n    </tr>\n    <tr>\n      <th>AL158050.2</th>\n      <td>Sudden cardiac arrest</td>\n    </tr>\n    <tr>\n      <th>LINC02376</th>\n      <td>Sudden cardiac arrest</td>\n    </tr>\n    <tr>\n      <th>AC073913.1</th>\n      <td>Sudden cardiac arrest</td>\n    </tr>\n    <tr>\n      <th>CHRNB4</th>\n      <td>Sudden cardiac arrest</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{},"execution_count":13}],"source":["df = df.explode('MAPPED_GENE')\n","df = df.set_index('MAPPED_GENE')\n","df.head()"]},{"cell_type":"code","execution_count":14,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"(253030, 1)"},"metadata":{},"execution_count":14}],"source":["df.shape"]},{"cell_type":"markdown","metadata":{},"source":[" # Filter Data"]},{"cell_type":"markdown","metadata":{},"source":[" ## Map Gene Symbols to Up-to-date Approved Gene Symbols"]},{"cell_type":"code","execution_count":91,"metadata":{},"outputs":[{"output_type":"stream","name":"stderr","text":"100%|██████████| 253030/253030 [00:00<00:00, 665767.13it/s]\n"},{"output_type":"execute_result","data":{"text/plain":"(175381, 1)"},"metadata":{},"execution_count":91}],"source":["df = uf.mapgenesymbols(df, symbol_lookup)\n","df.shape"]},{"cell_type":"markdown","metadata":{},"source":[" # Analyze Data"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Binary Matrix"]},{"cell_type":"code","execution_count":95,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"     \nA2M                                                     0          \nA2ML1                                                   0          \nA2MP1                                                   0          \nA4GALT                                                  0          \n\n        vascular endothelial growth factor D levels  \\\nA1CF                                              0   \nA2M                                               0   \nA2ML1                                             0   \nA2MP1                                             0   \nA4GALT                                            0   \n\n        visceral adipose tissue/total adipose tissue ratio  \nA1CF                                                    0   \nA2M                                                     0   \nA2ML1                                                   0   \nA2MP1                                                   0   \nA4GALT                                                  0   \n\n[5 rows x 4101 columns]","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>&amp;beta;2-Glycoprotein I (&amp;beta;2-GPI) plasma levels</th>\n      <th>1,5-anhydroglucitol levels</th>\n      <th>17-hydroxyprogesterone (17-OHP) levels</th>\n      <th>3-hydroxy-1-methylpropylmercapturic acid levels in smokers</th>\n      <th>3-hydroxypropylmercapturic acid levels in smokers</th>\n      <th>3-month functional outcome in ischaemic stroke (modified Rankin score)</th>\n      <th>3-month functional outcome in lacunar ischaemic stroke (modified Rankin score)</th>\n      <th>3-month functional outcome in non-lacunar ischaemic stroke (modified Rankin score)</th>\n      <th>5-HTT brain serotonin transporter levels</th>\n      <th>6-month creatinine clearance change response to tenofovir treatment in HIV infection (treatment arm interaction)</th>\n      <th>...</th>\n      <th>lower facial morphology traits (quantitative measurement)</th>\n      <th>melanoma-derived growth regulatory protein levels</th>\n      <th>middle facial morphology traits (quantitative measurement)</th>\n      <th>perceptual and visual search speed (trail making test A) (age interaction)</th>\n      <th>same-sex sexual behaviour</th>\n      <th>vWF and FVIII levels</th>\n      <th>vWF levels</th>\n      <th>vWF levels in ischaemic stroke and hyperhomocysteinaemia</th>\n      <th>vascular endothelial growth factor D levels</th>\n      <th>visceral adipose tissue/total adipose tissue ratio</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>A1CF</th>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>A2M</th>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>A2ML1</th>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>A2MP1</th>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>A4GALT</th>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n  </tbody>\n</table>\n<p>5 rows × 4101 columns</p>\n</div>"},"metadata":{},"execution_count":95}],"source":["binary_matrix = uf.createBinaryMatrix(df)\n","binary_matrix.head()"]},{"cell_type":"code","execution_count":96,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"(17954, 4101)"},"metadata":{},"execution_count":96}],"source":["binary_matrix.shape"]},{"cell_type":"code","execution_count":97,"metadata":{},"outputs":[],"source":["uf.saveData(binary_matrix, path, output_name + '_binary_matrix', \n","            compression='npz', dtype=np.uint8)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Gene List"]},{"cell_type":"code","execution_count":98,"metadata":{},"outputs":[{"output_type":"stream","name":"stderr","text":"100%|██████████| 17954/17954 [00:00<00:00, 188258.04it/s]\n"},{"output_type":"execute_result","data":{"text/plain":"  GeneSym  GeneID\n0    A1CF   29974\n1     A2M       2\n2   A2ML1  144568\n3   A2MP1       3\n4  A4GALT   53947","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>GeneSym</th>\n      <th>GeneID</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>A1CF</td>\n      <td>29974</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>A2M</td>\n      <td>2</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>A2ML1</td>\n      <td>144568</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>A2MP1</td>\n      <td>3</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>A4GALT</td>\n      <td>53947</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{},"execution_count":98}],"source":["gene_list = uf.createGeneList(binary_matrix, geneid_lookup)\n","gene_list.head()"]},{"cell_type":"code","execution_count":99,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"(17954, 2)"},"metadata":{},"execution_count":99}],"source":["gene_list.shape"]},{"cell_type":"code","execution_count":100,"metadata":{},"outputs":[],"source":["uf.saveData(gene_list, path, output_name + '_gene_list',\n","            ext='tsv', compression='gzip', index=False)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Attribute List"]},{"cell_type":"code","execution_count":101,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"Empty DataFrame\nColumns: []\nIndex: [&beta;2-Glycoprotein I (&beta;2-GPI) plasma levels, 1,5-anhydroglucitol levels, 17-hydroxyprogesterone (17-OHP) levels, 3-hydroxy-1-methylpropylmercapturic acid levels in smokers, 3-hydroxypropylmercapturic acid levels in smokers]","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n    </tr>\n    <tr>\n      <th>Attributes</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>&amp;beta;2-Glycoprotein I (&amp;beta;2-GPI) plasma levels</th>\n    </tr>\n    <tr>\n      <th>1,5-anhydroglucitol levels</th>\n    </tr>\n    <tr>\n      <th>17-hydroxyprogesterone (17-OHP) levels</th>\n    </tr>\n    <tr>\n      <th>3-hydroxy-1-methylpropylmercapturic acid levels in smokers</th>\n    </tr>\n    <tr>\n      <th>3-hydroxypropylmercapturic acid levels in smokers</th>\n    </tr>\n  </tbody>\n</table>\n</div>"},"metadata":{},"execution_count":101}],"source":["attribute_list = uf.createAttributeList(binary_matrix)\n","attribute_list.head()"]},{"cell_type":"code","execution_count":102,"metadata":{},"outputs":[{"output_type":"execute_result","data":{"text/plain":"(4101, 0)"},"metadata":{},"execution_count":102}],"source":["attribute_list.shape"]},{"cell_type":"code","execution_count":103,"metadata":{},"outputs":[],"source":["uf.saveData(attribute_list, path, output_name + '_attribute_list',\n","            ext='tsv', compression='gzip')"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Gene and Attribute Set Libraries"]},{"cell_type":"code","execution_count":104,"metadata":{},"outputs":[{"output_type":"stream","name":"stderr","text":"100%|██████████| 2135/2135 [00:00<00:00, 17208.93it/s]\n"}],"source":["uf.createUpGeneSetLib(binary_matrix, path, output_name + '_gene_up_set')"]},{"cell_type":"code","execution_count":105,"metadata":{},"outputs":[{"output_type":"stream","name":"stderr","text":"100%|██████████| 17954/17954 [00:00<00:00, 34317.98it/s]\n"}],"source":["uf.createUpAttributeSetLib(binary_matrix, path, \n","                           output_name + '_attribute_up_set')"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Attribute Similarity Matrix"]},{"cell_type":"code","execution_count":106,"metadata":{},"outputs":[],"source":["attribute_similarity_matrix = uf.createSimilarityMatrix(\n","                                binary_matrix.T, 'jaccard', sparse=True)\n","attribute_similarity_matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(attribute_similarity_matrix, path,\n","            output_name + '_attribute_similarity_matrix', \n","            compression='npz', symmetric=True, dtype=np.float32)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# net.load_df(attribute_similarity_matrix.iloc[:,:].copy())\n","# net.filter_N_top('row', rank_type='sum', N_top=300)\n","# net.cluster()\n","# net.widget()"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Gene Similarity Matrix"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["gene_similarity_matrix = uf.createSimilarityMatrix(binary_matrix, 'jaccard', sparse=True)\n","gene_similarity_matrix.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.saveData(gene_similarity_matrix, path, \n","            output_name + '_gene_similarity_matrix',\n","            compression='npz', symmetric=True, dtype=np.float32)"]},{"cell_type":"markdown","metadata":{},"source":[" ## Create Gene-Attribute Edge List"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.createGeneAttributeEdgeList(binary_matrix, attribute_list, gene_list, \n","                               path, output_name + '_gene_attribute_edge_list')"]},{"cell_type":"markdown","metadata":{},"source":[" # Create Downloadable Save File"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uf.createArchive(path)"]},{"cell_type":"markdown","metadata":{},"source":[" ### Link to download output files: [click here](./output_archive.zip)"]}],"metadata":{"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.0-final"},"orig_nbformat":2,"kernelspec":{"name":"python38064bitharmonizomevenve1fd62c035f347dbaffce382c9034785","display_name":"Python 3.8.0 64-bit ('harmonizome': venv)"}},"nbformat":4,"nbformat_minor":2}